File: scripts/rust/src/review_trait_definition_order.py
#!/usr/bin/env python3
"""
Review: Trait definition order - traits should appear before impl blocks.

Malpattern:
1. Data structure (struct/enum)
2. Inherent impl or trait impl <- WRONG: impl before trait definition
3. Trait definition for that data structure <- should be earlier

Correct order:
1. Data structure (struct/enum)
2. Trait definition for that data structure <- HERE
3. Inherent impl (impl Type { ... })
4. Custom trait impls
5. Standard trait impls

This ensures the trait interface is visible before any implementations.
"""
# Git commit: e4850e1
# Date: 2025-10-17

import re
import sys
from pathlib import Path


def is_struct_or_enum_line(line):
    """Check if line defines a struct or enum."""
    stripped = line.strip()
    return bool(re.match(r'(pub\s+)?(?:struct|enum)\s+\w+', stripped))


def is_trait_definition(line):
    """Check if line defines a trait."""
    stripped = line.strip()
    return bool(re.match(r'(pub\s+)?trait\s+\w+', stripped))


def is_impl_line(line):
    """Check if line starts an impl block."""
    stripped = line.strip()
    return bool(re.match(r'impl(?:<[^>]+>)?\s+', stripped))


def check_trait_order(file_path):
    """
    Check for trait definitions appearing after impl blocks.
    
    Returns list of violations: (file_path, struct_name, trait_line, trait_name, first_impl_line)
    """
    # Skip Types.rs - it has a different format
    if file_path.name == 'Types.rs':
        return []
    
    # Skip Chap47 (Claude abomination - will be replaced by Chap47clean)
    # Skip Chap47clean (different structure - needs interactive fixing)
    if 'Chap47' in str(file_path.parent):
        return []
    
    violations = []
    
    try:
        with open(file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
    except Exception as e:
        print(f"Error reading {file_path}: {e}", file=sys.stderr)
        return violations
    
    struct_name = None
    struct_line = None
    seen_impl_after_struct = False
    first_impl_line = None
    
    for i, line in enumerate(lines, 1):
        stripped = line.strip()
        
        # Skip empty lines and comments
        if not stripped or stripped.startswith('//'):
            continue
        
        # Detect struct/enum - resets state
        if is_struct_or_enum_line(line):
            m = re.search(r'(?:struct|enum)\s+(\w+)', stripped)
            struct_name = m.group(1) if m else None
            struct_line = i
            seen_impl_after_struct = False
            first_impl_line = None
            continue
        
        # Detect impl block (any kind)
        if struct_name and is_impl_line(line):
            if not seen_impl_after_struct:
                seen_impl_after_struct = True
                first_impl_line = i
            continue
        
        # Detect trait definition after impl
        if struct_name and seen_impl_after_struct and is_trait_definition(line):
            m = re.search(r'trait\s+(\w+)', stripped)
            trait_name = m.group(1) if m else 'Unknown'
            violations.append({
                'file': file_path,
                'struct': struct_name,
                'struct_line': struct_line,
                'trait_name': trait_name,
                'trait_line': i,
                'first_impl_line': first_impl_line,
            })
    
    return violations


def main():
    repo_root = Path(__file__).parent.parent.parent.parent
    
    search_dirs = [
        repo_root / "src",
    ]
    
    all_violations = []
    
    for search_dir in search_dirs:
        if not search_dir.exists():
            continue
        
        for rs_file in search_dir.rglob("*.rs"):
            violations = check_trait_order(rs_file)
            all_violations.extend(violations)
    
    if all_violations:
        print("✗ Trait Definition Order Violations:\n")
        print("Trait definitions should appear BEFORE impl blocks (after struct/enum).\n")
        
        for v in all_violations:
            rel_path = v['file'].relative_to(repo_root)
            print(f"  {rel_path}:{v['struct_line']}")
            print(f"    Struct: {v['struct']}")
            print(f"    Line {v['first_impl_line']}: First impl block")
            print(f"    Line {v['trait_line']}: trait {v['trait_name']} definition")
            print(f"    → Trait {v['trait_name']} should move before line {v['first_impl_line']}")
            print()
        
        print(f"Total violations: {len(all_violations)}")
        print("\nCorrect order:")
        print("  1. Data structure (struct/enum)")
        print("  2. Trait definition <- SHOULD BE HERE")
        print("  3. Inherent impl (impl Type { ... })")
        print("  4. Custom trait implementations")
        print("  5. Standard trait implementations")
        return 1
    
    print("✓ All trait definitions are in correct order")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/bin/review_trait_definition_order.rs
//! Review: Trait Definition Order
//!
//! Ensures trait definitions appear BEFORE impl blocks.
//!
//! Correct order:
//! 1. Data structure (struct/enum)
//! 2. Trait definition <- SHOULD BE HERE
//! 3. Inherent impl (impl Type { ... })
//! 4. Custom trait implementations
//! 5. Standard trait implementations
//!
//! Binary: rusticate-review-trait-definition-order

use anyhow::Result;
use ra_ap_syntax::{ast::{self, AstNode, HasName}, SyntaxKind, SourceFile, Edition};
use rusticate::{StandardArgs, find_rust_files};
use std::path::Path;
use std::time::Instant;
use std::fs;


macro_rules! log {
    ($($arg:tt)*) => {{
        use std::io::Write;
        let msg = format!($($arg)*);
        println!("{}", msg);
        if let Ok(mut file) = std::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open("analyses/review_trait_definition_order.log")
        {
            let _ = writeln!(file, "{}", msg);
        }
    }};
}
#[derive(Debug)]
struct Violation {
    struct_name: String,
    struct_line: usize,
    trait_name: String,
    trait_line: usize,
    first_impl_line: usize,
}

fn check_file(file_path: &Path) -> Vec<Violation> {
    // Skip specific files/dirs
    if file_path.file_name().map(|n| n == "Types.rs").unwrap_or(false) {
        return Vec::new();
    }
    if file_path.to_str().map(|s| s.contains("Chap47")).unwrap_or(false) {
        return Vec::new();
    }

    let content = match std::fs::read_to_string(file_path) {
        Ok(c) => c,
        Err(_) => return Vec::new(),
    };

    let parsed = SourceFile::parse(&content, Edition::Edition2021);
    let tree = parsed.tree();
    let root = tree.syntax();

    let mut violations = Vec::new();
    
    // Track state: struct name, first impl line, seen impl
    let mut current_struct: Option<(String, usize)> = None;
    let mut seen_impl_after_struct = false;
    let mut first_impl_line: Option<usize> = None;

    // Process top-level items in order
    for node in root.children() {
        let kind = node.kind();
        let line_num = content[..node.text_range().start().into()]
            .chars()
            .filter(|&c| c == '\n')
            .count() + 1;

        match kind {
            SyntaxKind::STRUCT | SyntaxKind::ENUM => {
                // Reset state for new struct/enum
                if let Some(adt) = ast::Struct::cast(node.clone())
                    .map(|s| s.name().map(|n| n.to_string()))
                    .flatten()
                    .or_else(|| {
                        ast::Enum::cast(node.clone())
                            .and_then(|e| e.name().map(|n| n.to_string()))
                    })
                {
                    current_struct = Some((adt, line_num));
                    seen_impl_after_struct = false;
                    first_impl_line = None;
                }
            }
            SyntaxKind::IMPL => {
                // Mark that we've seen an impl after current struct
                if current_struct.is_some() && !seen_impl_after_struct {
                    seen_impl_after_struct = true;
                    first_impl_line = Some(line_num);
                }
            }
            SyntaxKind::TRAIT => {
                // Check if trait appears after impl
                if let (Some((struct_name, struct_line)), Some(impl_line)) =
                    (&current_struct, first_impl_line)
                {
                    if seen_impl_after_struct {
                        if let Some(trait_node) = ast::Trait::cast(node.clone()) {
                            if let Some(trait_name) = trait_node.name() {
                                violations.push(Violation {
                                    struct_name: struct_name.clone(),
                                    struct_line: *struct_line,
                                    trait_name: trait_name.to_string(),
                                    trait_line: line_num,
                                    first_impl_line: impl_line,
                                });
                            }
                        }
                    }
                }
            }
            _ => {}
        }
    }

    violations
}

fn main() -> Result<()> {
    let start_time = Instant::now();
    let args = StandardArgs::parse()?;
    let base_dir = args.base_dir();

    let files = find_rust_files(&args.paths);

    let mut all_violations = Vec::new();

    for file_path in &files {
        let violations = check_file(file_path);
        for violation in violations {
            let rel_path = file_path.strip_prefix(&base_dir).unwrap_or(file_path);
            all_violations.push((rel_path.to_path_buf(), violation));
        }
    }

    if all_violations.is_empty() {
        log!("✓ All trait definitions are in correct order");
        let elapsed = start_time.elapsed();
        eprintln!("\nCompleted in {}ms", elapsed.as_millis());
        return Ok(());
    }

    log!("✗ Trait Definition Order Violations:\n");
    log!("Trait definitions should appear BEFORE impl blocks (after struct/enum).\n");

    for (file_path, v) in &all_violations {
        log!("  {}:{}", file_path.display(), v.struct_line);
        log!("    Struct: {}", v.struct_name);
        log!("    Line {}: First impl block", v.first_impl_line);
        log!("    Line {}: trait {} definition", v.trait_line, v.trait_name);
        log!("    → Trait {} should move before line {}", v.trait_name, v.first_impl_line);
        log!("");
    }

    log!("Total violations: {}", all_violations.len());
    log!("\nCorrect order:");
    log!("  1. Data structure (struct/enum)");
    log!("  2. Trait definition <- SHOULD BE HERE");
    log!("  3. Inherent impl (impl Type {{ ... }})");
    log!("  4. Custom trait implementations");
    log!("  5. Standard trait implementations");

    let elapsed = start_time.elapsed();
    eprintln!("\nCompleted in {}ms", elapsed.as_millis());

    Ok(())
}

